File: jar/src/main/java/jline/opt/solver/de/DifferentialEvolution.java
package jline.opt.solver.de;

import java.util.ArrayList;
import java.util.List;

/**
 * Self-contained port of scipy's {@code differential_evolution} that reproduces
 * its trajectory bit-for-bit for a given integer seed, using
 * {@link NumpyRandomState} for all draws.
 *
 * <p>Covers the configuration used by line-opt: the binomial strategies
 * (default {@code best1bin}), {@code updating='immediate'}, dithered mutation,
 * latinhypercube initialization, {@code polish=false}, {@code workers=1}, and
 * penalty-based constraints handled inside the objective (no scipy-level
 * constraints). The exact numpy draw order is preserved: LHS init
 * ({@code uniform} grid then per-column {@code permutation}), a per-generation
 * dither {@code uniform}, and per candidate {@code randint} (fill point),
 * {@code shuffle} (sample selection) and {@code uniform} (crossover), plus a
 * data-dependent {@code uniform} in the bound-repair step.</p>
 *
 * <p>Java 8 compatible.</p>
 */
public class DifferentialEvolution {

    /** Objective returning the (penalized) energy; may be {@code +inf}. */
    public interface Objective {
        double evaluate(double[] x);
    }

    /** Per-generation callback; return true to stop early. */
    public interface Callback {
        boolean call(double[] bestX, int nit);
    }

    public static class Result {
        public double[] x;
        public double fun;
        public int nit;
        public int nfev;
        public boolean success;
        public List<double[]> bestPerGen = new ArrayList<double[]>();
    }

    private final Objective objective;
    private final double[] low;
    private final double[] high;
    private final int paramCount;

    private final String strategy;
    private final int popsizeMult;
    private final int maxiter;
    private final double ditherLow;
    private final double ditherHigh;
    private final double recombination;
    private final double tol;
    private final double atol;
    private Callback callback;

    private final NumpyRandomState rng;

    private int numMembers;
    private double[][] population;      // [member][param], scaled to [0,1]
    private double[] energies;
    private int[] randomIndex;          // persistent shuffle buffer
    private double scale;
    private int nfev;

    private static final double MACHEPS = Math.ulp(1.0);

    public DifferentialEvolution(Objective objective, double[] low, double[] high,
                                 String strategy, int popsizeMult, int maxiter,
                                 double ditherLow, double ditherHigh,
                                 double recombination, double tol, long seed) {
        this.objective = objective;
        this.low = low.clone();
        this.high = high.clone();
        this.paramCount = low.length;
        this.strategy = strategy;
        this.popsizeMult = popsizeMult;
        this.maxiter = maxiter;
        this.ditherLow = Math.min(ditherLow, ditherHigh);
        this.ditherHigh = Math.max(ditherLow, ditherHigh);
        this.recombination = recombination;
        this.tol = tol;
        this.atol = 0.0;
        this.scale = ditherLow;
        this.rng = new NumpyRandomState(seed);
        initPopulationLhs();
    }

    public void setCallback(Callback cb) {
        this.callback = cb;
    }

    public NumpyRandomState getRng() {
        return rng;
    }

    // ---- scaling between population [0,1] and parameter space --------------

    private double scaleArg1(int j) {
        return 0.5 * (low[j] + high[j]);
    }

    private double scaleArg2(int j) {
        return Math.abs(low[j] - high[j]);
    }

    private double[] scaleParameters(double[] trial) {
        double[] out = new double[paramCount];
        for (int j = 0; j < paramCount; j++) {
            out[j] = scaleArg1(j) + (trial[j] - 0.5) * scaleArg2(j);
        }
        return out;
    }

    // ---- initialization ----------------------------------------------------

    private void initPopulationLhs() {
        int ebCount = 0;
        for (int j = 0; j < paramCount; j++) {
            if (low[j] == high[j]) {
                ebCount++;
            }
        }
        numMembers = Math.max(5, popsizeMult * Math.max(1, paramCount - ebCount));

        double segsize = 1.0 / numMembers;
        // samples = segsize * uniform(size=(numMembers, paramCount))
        //           + linspace(0,1,numMembers,endpoint=False)[:,None]
        double[][] samples = new double[numMembers][paramCount];
        double[] flat = rng.uniform(numMembers * paramCount);
        int p = 0;
        for (int i = 0; i < numMembers; i++) {
            double offset = ((double) i) / numMembers;
            for (int j = 0; j < paramCount; j++) {
                samples[i][j] = segsize * flat[p++] + offset;
            }
        }
        population = new double[numMembers][paramCount];
        for (int j = 0; j < paramCount; j++) {
            int[] order = rng.permutation(numMembers);
            for (int i = 0; i < numMembers; i++) {
                population[i][j] = samples[order[i]][j];
            }
        }
        energies = new double[numMembers];
        for (int i = 0; i < numMembers; i++) {
            energies[i] = Double.POSITIVE_INFINITY;
        }
        randomIndex = new int[numMembers];
        for (int i = 0; i < numMembers; i++) {
            randomIndex[i] = i;
        }
        nfev = 0;
    }

    // ---- helpers -----------------------------------------------------------

    private double[] best() {
        return scaleParameters(population[0]);
    }

    private void calculateInitialEnergies() {
        for (int i = 0; i < numMembers; i++) {
            energies[i] = objective.evaluate(scaleParameters(population[i]));
            nfev++;
        }
    }

    private void promoteLowestEnergy() {
        int l = 0;
        double best = energies[0];
        for (int i = 1; i < numMembers; i++) {
            if (energies[i] < best) {
                best = energies[i];
                l = i;
            }
        }
        if (l != 0) {
            double te = energies[0];
            energies[0] = energies[l];
            energies[l] = te;
            double[] tp = population[0];
            population[0] = population[l];
            population[l] = tp;
        }
    }

    private double std(double[] a) {
        double mean = 0.0;
        for (double v : a) {
            mean += v;
        }
        mean /= a.length;
        double var = 0.0;
        for (double v : a) {
            var += (v - mean) * (v - mean);
        }
        return Math.sqrt(var / a.length);
    }

    private double mean(double[] a) {
        double m = 0.0;
        for (double v : a) {
            m += v;
        }
        return m / a.length;
    }

    private boolean anyInf(double[] a) {
        for (double v : a) {
            if (Double.isInfinite(v)) {
                return true;
            }
        }
        return false;
    }

    private boolean converged() {
        if (anyInf(energies)) {
            return false;
        }
        return std(energies) <= atol + tol * Math.abs(mean(energies));
    }

    // ---- sample selection and mutation ------------------------------------

    private int[] selectSamples(int candidate, int numberSamples) {
        rng.shuffle(randomIndex);
        int[] pick = new int[numberSamples + 1];
        System.arraycopy(randomIndex, 0, pick, 0, numberSamples + 1);
        int[] out = new int[numberSamples];
        int k = 0;
        for (int i = 0; i < pick.length && k < numberSamples; i++) {
            if (pick[i] != candidate) {
                out[k++] = pick[i];
            }
        }
        return out;
    }

    private double[] bprime(int candidate, int[] s) {
        double[] b = new double[paramCount];
        if ("rand1bin".equals(strategy) || "rand1exp".equals(strategy)) {
            for (int j = 0; j < paramCount; j++) {
                b[j] = population[s[0]][j] + scale
                        * (population[s[1]][j] - population[s[2]][j]);
            }
        } else if ("randtobest1bin".equals(strategy) || "randtobest1exp".equals(strategy)) {
            for (int j = 0; j < paramCount; j++) {
                double v = population[s[0]][j];
                v += scale * (population[0][j] - v);
                v += scale * (population[s[1]][j] - population[s[2]][j]);
                b[j] = v;
            }
        } else if ("currenttobest1bin".equals(strategy) || "currenttobest1exp".equals(strategy)) {
            for (int j = 0; j < paramCount; j++) {
                b[j] = population[candidate][j] + scale
                        * (population[0][j] - population[candidate][j]
                           + population[s[0]][j] - population[s[1]][j]);
            }
        } else if ("best2bin".equals(strategy) || "best2exp".equals(strategy)) {
            for (int j = 0; j < paramCount; j++) {
                b[j] = population[0][j] + scale
                        * (population[s[0]][j] + population[s[1]][j]
                           - population[s[2]][j] - population[s[3]][j]);
            }
        } else if ("rand2bin".equals(strategy) || "rand2exp".equals(strategy)) {
            for (int j = 0; j < paramCount; j++) {
                b[j] = population[s[0]][j] + scale
                        * (population[s[1]][j] + population[s[2]][j]
                           - population[s[3]][j] - population[s[4]][j]);
            }
        } else {
            // default best1bin / best1exp
            for (int j = 0; j < paramCount; j++) {
                b[j] = population[0][j] + scale
                        * (population[s[0]][j] - population[s[1]][j]);
            }
        }
        return b;
    }

    private double[] mutate(int candidate) {
        int fillPoint = (int) rng.randint(0, paramCount);
        int[] samples = selectSamples(candidate, 5);
        double[] bp = bprime(candidate, samples);
        double[] trial = population[candidate].clone();
        double[] cross = rng.uniform(paramCount);
        boolean[] doCross = new boolean[paramCount];
        for (int j = 0; j < paramCount; j++) {
            doCross[j] = cross[j] < recombination;
        }
        // binomial: force the fill point
        doCross[fillPoint] = true;
        for (int j = 0; j < paramCount; j++) {
            if (doCross[j]) {
                trial[j] = bp[j];
            }
        }
        return trial;
    }

    private void ensureConstraint(double[] trial) {
        int oob = 0;
        for (int j = 0; j < paramCount; j++) {
            if (trial[j] > 1 || trial[j] < 0) {
                oob++;
            }
        }
        if (oob > 0) {
            double[] repl = rng.uniform(oob);
            int k = 0;
            for (int j = 0; j < paramCount; j++) {
                if (trial[j] > 1 || trial[j] < 0) {
                    trial[j] = repl[k++];
                }
            }
        }
    }

    // ---- main loop ---------------------------------------------------------

    public Result solve() {
        Result result = new Result();
        boolean warningFlag = false;

        if (anyInf(energies)) {
            calculateInitialEnergies();
            promoteLowestEnergy();
        }

        int nit = 0;
        for (nit = 1; nit <= maxiter; nit++) {
            next();

            if (callback != null) {
                double[] bx = best();
                boolean stop = callback.call(bx, nit);
                if (stop) {
                    warningFlag = true;
                }
            }
            result.bestPerGen.add(best());

            if (warningFlag || converged()) {
                break;
            }
        }
        if (nit > maxiter) {
            nit = maxiter;
            warningFlag = true;
        }

        result.x = best();
        result.fun = energies[0];
        result.nit = nit;
        result.nfev = nfev;
        result.success = !warningFlag;
        return result;
    }

    private void next() {
        // dither each generation (mutation is a tuple -> dither active)
        scale = rng.uniform(ditherLow, ditherHigh);

        for (int candidate = 0; candidate < numMembers; candidate++) {
            double[] trial = mutate(candidate);
            ensureConstraint(trial);
            double[] parameters = scaleParameters(trial);
            double energy = objective.evaluate(parameters);
            nfev++;
            if (energy <= energies[candidate]) {
                population[candidate] = trial;
                energies[candidate] = energy;
                if (energy <= energies[0]) {
                    promoteLowestEnergy();
                }
            }
        }
    }
}

File: matlab/src/opt/+opt/+de/DifferentialEvolution.m
classdef DifferentialEvolution < handle
    % DifferentialEvolution  Self-contained port of scipy's
    % differential_evolution that reproduces its trajectory bit-for-bit for a
    % given integer seed, using opt.de.NumpyRandomState for all draws.
    %
    % Covers the configuration used by line-opt: binomial strategies (default
    % best1bin), updating='immediate', dithered mutation, latinhypercube
    % initialization, polish=false, and penalty-based constraints handled inside
    % the objective. The exact numpy draw order is preserved: LHS init (uniform
    % grid then per-column permutation), a per-generation dither uniform, and per
    % candidate randint (fill point), shuffle (sample selection) and uniform
    % (crossover), plus a data-dependent uniform in the bound-repair step.

    properties
        objective          % function handle: energy = f(x)
        low                % 1xd lower bounds
        high               % 1xd upper bounds
        paramCount
        strategy
        popsizeMult
        maxiter
        ditherLow
        ditherHigh
        recombination
        tol
        atol
        callback           % function handle (bestX, nit) -> logical stop, or []
        rng                % opt.de.NumpyRandomState
        numMembers
        population         % numMembers x d, scaled to [0,1]
        energies           % 1 x numMembers
        randomIndex        % persistent shuffle buffer (0-based values)
        scale
        nfev
        bestPerGen         % cell array of 1xd vectors
    end

    methods
        function obj = DifferentialEvolution(objective, low, high, strategy, ...
                popsizeMult, maxiter, ditherLow, ditherHigh, recombination, tol, seed)
            obj.objective = objective;
            obj.low = low(:).';
            obj.high = high(:).';
            obj.paramCount = numel(low);
            obj.strategy = strategy;
            obj.popsizeMult = popsizeMult;
            obj.maxiter = maxiter;
            obj.ditherLow = min(ditherLow, ditherHigh);
            obj.ditherHigh = max(ditherLow, ditherHigh);
            obj.recombination = recombination;
            obj.tol = tol;
            obj.atol = 0.0;
            obj.callback = [];
            obj.scale = ditherLow;
            obj.rng = opt.de.NumpyRandomState(seed);
            obj.bestPerGen = {};
            obj.initPopulationLhs();
        end

        function s1 = scaleArg1(obj, j)
            s1 = 0.5 * (obj.low(j) + obj.high(j));
        end

        function s2 = scaleArg2(obj, j)
            s2 = abs(obj.low(j) - obj.high(j));
        end

        function out = scaleParameters(obj, trial)
            out = zeros(1, obj.paramCount);
            for j = 1:obj.paramCount
                out(j) = obj.scaleArg1(j) + (trial(j) - 0.5) * obj.scaleArg2(j);
            end
        end

        function initPopulationLhs(obj)
            ebCount = sum(obj.low == obj.high);
            obj.numMembers = max(5, obj.popsizeMult * max(1, obj.paramCount - ebCount));
            segsize = 1.0 / obj.numMembers;
            flat = obj.rng.uniformN(0, 1, obj.numMembers * obj.paramCount);
            samples = zeros(obj.numMembers, obj.paramCount);
            p = 1;
            for i = 1:obj.numMembers
                offset = (i - 1) / obj.numMembers;
                for j = 1:obj.paramCount
                    samples(i, j) = segsize * flat(p) + offset;
                    p = p + 1;
                end
            end
            obj.population = zeros(obj.numMembers, obj.paramCount);
            for j = 1:obj.paramCount
                order = obj.rng.permutation(obj.numMembers);  % 0-based
                for i = 1:obj.numMembers
                    obj.population(i, j) = samples(order(i) + 1, j);
                end
            end
            obj.energies = inf(1, obj.numMembers);
            obj.randomIndex = 0:(obj.numMembers - 1);
            obj.nfev = 0;
        end

        function b = best(obj)
            b = obj.scaleParameters(obj.population(1, :));
        end

        function calculateInitialEnergies(obj)
            for i = 1:obj.numMembers
                obj.energies(i) = obj.objective(obj.scaleParameters(obj.population(i, :)));
                obj.nfev = obj.nfev + 1;
            end
        end

        function promoteLowestEnergy(obj)
            [~, l] = min(obj.energies);
            if l ~= 1
                te = obj.energies(1); obj.energies(1) = obj.energies(l); obj.energies(l) = te;
                tp = obj.population(1, :); obj.population(1, :) = obj.population(l, :); obj.population(l, :) = tp;
            end
        end

        function tf = converged(obj)
            if any(isinf(obj.energies))
                tf = false;
                return;
            end
            tf = std(obj.energies, 1) <= obj.atol + obj.tol * abs(mean(obj.energies));
        end

        function out = selectSamples(obj, candidate, numberSamples)
            % candidate is 0-based member index
            obj.randomIndex = obj.rng.shuffle(obj.randomIndex);
            pick = obj.randomIndex(1:numberSamples + 1);
            out = zeros(1, numberSamples);
            k = 0;
            for i = 1:numel(pick)
                if k >= numberSamples
                    break;
                end
                if pick(i) ~= candidate
                    k = k + 1;
                    out(k) = pick(i);
                end
            end
        end

        function b = bprime(obj, candidate, s)
            % candidate and s are 0-based member indices; convert to 1-based rows
            P = obj.population;
            c = candidate + 1;
            r = s + 1;
            switch obj.strategy
                case {'rand1bin', 'rand1exp'}
                    b = P(r(1), :) + obj.scale * (P(r(2), :) - P(r(3), :));
                case {'randtobest1bin', 'randtobest1exp'}
                    b = P(r(1), :);
                    b = b + obj.scale * (P(1, :) - b);
                    b = b + obj.scale * (P(r(2), :) - P(r(3), :));
                case {'currenttobest1bin', 'currenttobest1exp'}
                    b = P(c, :) + obj.scale * (P(1, :) - P(c, :) + P(r(1), :) - P(r(2), :));
                case {'best2bin', 'best2exp'}
                    b = P(1, :) + obj.scale * (P(r(1), :) + P(r(2), :) - P(r(3), :) - P(r(4), :));
                case {'rand2bin', 'rand2exp'}
                    b = P(r(1), :) + obj.scale * (P(r(2), :) + P(r(3), :) - P(r(4), :) - P(r(5), :));
                otherwise   % best1bin / best1exp
                    b = P(1, :) + obj.scale * (P(r(1), :) - P(r(2), :));
            end
        end

        function trial = mutate(obj, candidate)
            % candidate is 0-based member index
            fillPoint = obj.rng.randint(0, obj.paramCount);   % 0-based
            samples = obj.selectSamples(candidate, 5);
            bp = obj.bprime(candidate, samples);
            trial = obj.population(candidate + 1, :);
            cross = obj.rng.uniformN(0, 1, obj.paramCount);
            doCross = cross < obj.recombination;
            doCross(fillPoint + 1) = true;
            trial(doCross) = bp(doCross);
        end

        function trial = ensureConstraint(obj, trial)
            mask = (trial > 1) | (trial < 0);
            oob = sum(mask);
            if oob > 0
                repl = obj.rng.uniformN(0, 1, oob);
                trial(mask) = repl;
            end
        end

        function result = solve(obj)
            warningFlag = false;
            if any(isinf(obj.energies))
                obj.calculateInitialEnergies();
                obj.promoteLowestEnergy();
            end
            nit = 0;
            for nit = 1:obj.maxiter
                obj.next();
                if ~isempty(obj.callback)
                    stop = obj.callback(obj.best(), nit);
                    if stop
                        warningFlag = true;
                    end
                end
                obj.bestPerGen{end+1} = obj.best(); %#ok<AGROW>
                if warningFlag || obj.converged()
                    break;
                end
            end
            if nit >= obj.maxiter
                warningFlag = warningFlag || (nit == obj.maxiter && ~obj.converged());
            end
            result = struct();
            result.x = obj.best();
            result.fun = obj.energies(1);
            result.nit = nit;
            result.nfev = obj.nfev;
            result.success = ~warningFlag;
            result.bestPerGen = obj.bestPerGen;
        end

        function next(obj)
            obj.scale = obj.rng.uniformScalar(obj.ditherLow, obj.ditherHigh);
            for candidate = 0:(obj.numMembers - 1)
                trial = obj.mutate(candidate);
                trial = obj.ensureConstraint(trial);
                parameters = obj.scaleParameters(trial);
                energy = obj.objective(parameters);
                obj.nfev = obj.nfev + 1;
                if energy <= obj.energies(candidate + 1)
                    obj.population(candidate + 1, :) = trial;
                    obj.energies(candidate + 1) = energy;
                    if energy <= obj.energies(1)
                        obj.promoteLowestEnergy();
                    end
                end
            end
        end
    end
end
